File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
# class Solution(object):
#     def minEatingSpeed(self, piles, h):
#         """
#         :type piles: List[int]
#         :type h: int
#         :rtype: int
#         """
#         max1=0
#         for p in piles:
#             max1=max(max1,p)

#         def totalhours(v,h):
#             totalhour=0
#             for p in v:
#                 totalhour+=math.ceil(p/h)
#             return totalhour
        
#         l=1
#         r=max1
#         m=0
#         while l<=r:
#             m=(l+r)//2
#             # if totalhours(piles,m)<h:
#             #     r=m-1
#             if totalhours(piles,m)<=h:
#                 r=m-1
#             else:
#                 l=m+1
#             # elif totalhours(piles,m)==h:
#             #     l=r=m
#         return l

# class Solution:
#     def minEatingSpeed(self, piles, h):
        
#         l, r = 1, max(piles)
#         res = r
        
#         while l <= r:
#             k = (l+r)//2 
#             totalHours = 0
            
#             for i in piles:
#                 totalHours += math.ceil(i/k)
            
#             if totalHours <= h:
#                 res = min(res, k)
#                 r = k - 1
#             else:
#                 l = k + 1
        
#         return res
                
        
        
        
            
        
        
        
        
class Solution:
    def minEatingSpeed(self, piles, h):
        def check_can_eat(k):
            actual_h = 0
            for pile in piles:
                actual_h += (pile+k-1)//k
            return actual_h <= h
        l,r = 0,max(piles)
        while r-l>1:
            m=(l+r)//2
            if check_can_eat(m):
                r = m
            else:
                l = m
        return r        

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
import java.util.Arrays;

class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l = 1; // Minimum possible eating speed
        int r = findMax(piles); // Maximum possible eating speed
        int ans = r;
        while (l <= r) {
            int mid = l + (r - l) / 2; // Middle speed to check
            
            if (canEatAll(piles, mid, h)) {
                ans = mid;
                r = mid - 1; // Try for a smaller speed
            } else {
                l = mid + 1; // Increase speed
            }
        }
        
        return ans;
    }
    
    private boolean canEatAll(int[] piles, int k, int h) {
        int days = 0;
        
        // Calculate total days needed at speed k
        for (double pile : piles) {
            days += Math.ceil(pile / (double)(k));
        }
        
        return days <= h; // Return whether it fits within the allowed hours
    }
    
    private int findMax(int[] v) {
        int maxi = Integer.MIN_VALUE;;
        int n = v.length;
        //find the maximum:
        for (int i = 0; i < n; i++) {
            maxi = Math.max(maxi, v[i]);
        }
        return maxi;
    }
}
